File: priv/static_epmd/static_epmd.ex
defmodule StaticEpmd do
  # THIS MUST BE PRECOMPILED BEFORE STARTING THE SERVER
  # To compile this:
  # elixirc static_epmd.ex

  # These are the mandatory callbacks for the EPMD protocol.
  # We are implementing the 'erl_epmd' behavior.

  # 1. Start Link: We don't need a daemon process, so we ignore this.
  def start_link, do: :ignore

  # 2. Register Node: We don't register anything; we rely on static ports.
  # We return a random "creation" number (1-3) to satisfy the protocol.
  def register_node(name, port), do: register_node(name, port, :inet)
  def register_node(_name, _port, _driver), do: {:ok, :rand.uniform(3)}

  # 3. Names: Used for debugging (e.g., net_adm:names()). We return empty.
  def names(_host), do: {:ok, ["StaticNames"]}

  # 4. Port Please: The core logic.
  # 'name' comes in as a charlist (e.g. 'nodeA_9100'), NOT a string.
  # IMPORTANT: Must use pure Erlang functions, not Elixir stdlib
  def port_please(name, _ip) when is_list(name) do
    # Split charlist by underscore manually (95 is ASCII for '_')
    # e.g. 'local2_9200' -> ['local2', '9200']
    parts = split_charlist(name, ?_)

    result = case parts do
      [] ->
        :io.format("[StaticEpmd] ERROR: Empty name~n")
        :noport
      parts_list ->
        # Get the last part (should be the port)
        port_part = :lists.last(parts_list)
        parsed = parse_port_charlist(port_part)

        # Log for debugging
        :io.format("[StaticEpmd] Resolving '~s' parts=~p -> ~p~n", [name, parts, parsed])
        parsed
    end

    result
  end

  # Fallback for non-list input
  def port_please(name, _ip) do
    :io.format("[StaticEpmd] ERROR: Invalid name type: ~p~n", [name])
    :noport
  end

  # Manually split a charlist by a delimiter
  defp split_charlist(list, delimiter) do
    split_charlist(list, delimiter, [], [])
  end

  defp split_charlist([], _delimiter, current, acc) do
    # Add the last part and reverse to get correct order
    :lists.reverse([:lists.reverse(current) | acc])
  end

  defp split_charlist([char | rest], delimiter, current, acc) when char == delimiter do
    # Found delimiter, save current part and start new one
    split_charlist(rest, delimiter, [], [:lists.reverse(current) | acc])
  end

  defp split_charlist([char | rest], delimiter, current, acc) do
    # Regular character, add to current part
    split_charlist(rest, delimiter, [char | current], acc)
  end

  # Helper to parse the port charlist safely
  # Uses pure Erlang functions for compatibility
  defp parse_port_charlist(port_chars) when is_list(port_chars) do
    # Try to convert charlist to integer
    # e.g. '9200' -> 9200
    try do
      port = :erlang.list_to_integer(port_chars)
      {:port, port, 6}  # 6 is the protocol version for OTP 23+
    rescue
      ArgumentError ->
        :io.format("[StaticEpmd] ERROR: Invalid port format: ~p~n", [port_chars])
        :noport
    end
  end

  defp parse_port_charlist(_), do: :noport


  # 6. Address Please: Resolves hostname and returns IP + port
  def address_please(name, host, address_family) when is_list(name) do
    :io.format("[StaticEpmd] address_please called for: ~s@~s~n", [name, host])

    # First resolve the hostname to an IP
    case :inet.getaddr(host, address_family) do
      {:ok, ip} ->
        :io.format("[StaticEpmd] Resolved ~s to ~p~n", [host, ip])

        # Then get the port from the name
        parts = split_charlist(name, ?_)

        case parts do
          [] ->
            {:error, :nxdomain}
          parts_list ->
            port_part = :lists.last(parts_list)

            try do
              port = :erlang.list_to_integer(port_part)
              :io.format("[StaticEpmd] Returning: ip=~p port=~p version=6~n", [ip, port])
              {:ok, ip, port, 6}
            rescue
              ArgumentError ->
                {:error, :nxdomain}
            end
        end

      {:error, reason} ->
        :io.format("[StaticEpmd] Failed to resolve ~s: ~p~n", [host, reason])
        {:error, reason}
    end
  end

  def address_please(name, _host, _address_family) do
    :io.format("[StaticEpmd] ERROR: Invalid name type in address_please: ~p~n", [name])
    {:error, :nxdomain}
  end
end

File: priv/static_epmd/static_epmd.erl
-module(static_epmd).
-behaviour(gen_server).

%% This module implements a static EPMD (Erlang Port Mapper Daemon) replacement
%% that extracts port numbers from node names (e.g., node_9100 -> port 9100).
%%
%% Based on epmdless_static from https://github.com/tsloughter/epmdless
%% with added support for dynamic port detection from node names.
%%
%% To use this module, compile it and pass it to the Erlang VM:
%% -epmd_module static_epmd

%% API
-export([start_link/0,
         register_node/3,
         port_please/2,
         address_please/3,
         listen_port_please/2,
         names/1]).

%% gen_server callbacks
-export([init/1,
         handle_call/3,
         handle_cast/2,
         handle_info/2,
         terminate/2,
         code_change/3]).

-ifdef(OTP_RELEASE).
-if(?OTP_RELEASE >= 23).
-define(ERL_DIST_VER, 6).
-else.
-define(ERL_DIST_VER, 5).
-endif.
-else.
-define(ERL_DIST_VER, 5).
-endif.

-record(state, {port :: inet:port_number()}).

%%====================================================================
%% API
%%====================================================================

start_link() ->
    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).

%% Called when registering the local node
register_node(_Name, _Port, _Driver) ->
    %% Return a random creation number (required by protocol)
    {ok, rand:uniform(3)}.

%% Called to find the port for a remote node
%% This is where we extract the port from the node name
port_please(Name, _Ip) when is_list(Name) ->
    io:format("[StaticEpmd] port_please called for: ~s~n", [Name]),

    %% Try to extract port from node name (e.g., "node_9100" -> 9100)
    case extract_port_from_name(Name) of
        {ok, Port} ->
            io:format("[StaticEpmd] Extracted port ~p from name ~s~n", [Port, Name]),
            {port, Port, ?ERL_DIST_VER};
        error ->
            %% If we can't extract port from name, fall back to configured port
            io:format("[StaticEpmd] Could not extract port from name ~s, using default~n", [Name]),
            case gen_server:call(?MODULE, port_please, infinity) of
                {ok, Port} ->
                    {port, Port, ?ERL_DIST_VER};
                error ->
                    noport
            end
    end.

%% Called to resolve hostname to IP address
address_please(Name, Host, AddressFamily) ->
    io:format("[StaticEpmd] address_please called for: ~s@~s~n", [Name, Host]),

    %% First resolve the host to an IP
    case inet:getaddr(Host, AddressFamily) of
        {ok, Ip} ->
            %% Then get the port
            case port_please(Name, Ip) of
                {port, Port, Version} ->
                    {ok, Ip, Port, Version};
                noport ->
                    {error, nxdomain}
            end;
        {error, _} = Error ->
            Error
    end.

%% Called to get the local listening port
listen_port_please(_Name, _Host) ->
    case gen_server:call(?MODULE, port_please, infinity) of
        {ok, Port} ->
            {ok, Port};
        error ->
            {error, no_port_configured}
    end.

%% Called to list registered names (used by net_adm:names())
names(_Hostname) ->
    %% We don't maintain a registry, so return error
    {error, address}.

%%====================================================================
%% gen_server callbacks
%%====================================================================

init([]) ->
    %% Get the port from environment variable or command-line args
    Port = case os:getenv("ERL_DIST_PORT") of
        false ->
            %% Try command-line args (e.g., -erl_epmd_port 9100)
            case init:get_argument(erl_epmd_port) of
                {ok, [[PortStr]]} ->
                    list_to_integer(PortStr);
                _ ->
                    %% Default fallback
                    io:format("[StaticEpmd] WARNING: No ERL_DIST_PORT env var or -erl_epmd_port arg found~n"),
                    0
            end;
        PortStr when is_list(PortStr) ->
            list_to_integer(PortStr);
        PortStr ->
            binary_to_integer(list_to_binary(PortStr))
    end,

    io:format("[StaticEpmd] Initialized with local port: ~p~n", [Port]),
    {ok, #state{port = Port}}.

handle_call(port_please, _From, #state{port = 0} = State) ->
    %% No port configured
    {reply, error, State};
handle_call(port_please, _From, #state{port = Port} = State) ->
    {reply, {ok, Port}, State};
handle_call(_Request, _From, State) ->
    {reply, ok, State}.

handle_cast(_Msg, State) ->
    {noreply, State}.

handle_info(_Info, State) ->
    {noreply, State}.

terminate(_Reason, _State) ->
    ok.

code_change(_OldVsn, State, _Extra) ->
    {ok, State}.

%%====================================================================
%% Internal functions
%%====================================================================

%% Extract port number from node name
%% Node names are expected to be in format: name_PORT (e.g., "gateway_9100")
%% The port is always the last part after the last underscore
extract_port_from_name(Name) when is_list(Name) ->
    %% Split by underscore (95 is ASCII for '_')
    Parts = string:split(Name, "_", all),

    case Parts of
        [] ->
            error;
        PartsList ->
            %% Get the last part
            LastPart = lists:last(PartsList),

            %% Try to convert to integer
            try
                Port = list_to_integer(LastPart),
                %% Validate port range
                if
                    Port > 0 andalso Port =< 65535 ->
                        {ok, Port};
                    true ->
                        error
                end
            catch
                error:badarg ->
                    error
            end
    end;
extract_port_from_name(_) ->
    error.
